[FIX] tools.mail: ignore original email during encapsulation

odony opened the pull request:
When the system broadcasts an email response to document followers, if the config parameters `mail.force.smtp.from` or `mail.dynamic.smtp.from` are defined, it will rewrite the `From` address to avoid spoofing the sender's domain.

For example, if the `mail.catchall.domain` is set to `example.com` and an email response comes from:

   `"John D" <john@doe.com>`

it will rewrite it to:

   `"John D (john@doe.com)" <notifications@example.com>`

This will make sure the system never sends outgoing email for an external domain, as it has no authority for doing so, and that could break mail filtering/authentication rules (SPF, DMARC, etc.)

During this "encapsulation rewrite step", both the original Sender name and their email are preserved, and put into the quoted "name" field of the rewritten address. It seems sensible to preserve as much information as possible about the original sender.

Unfortunately, the inclusion of the Sender email in the final name makes it appear to some inbox providers as if the message is trying to deceptively impersonate another person (as many phishing schemes would).
As of November 2021 GMail at least does this, and will hide the name in the UI when it happens. It will keep only the rewritten email, which is not very useful in the case of a notification (even though it's more technically correct, of course).

This patch removes the original email from the rewritten notification, keeping only the name, considering that the email is not the most important part, and it's better to have one of the two than none.

So after the patch, the rewritten address is now:
    
  `"John D" <notifications@example.com>`
    
When there is no name in the original address, we keep only the local part of the email, to avoid the same display issue. The recipient will have to identify the sender based on the context / past messages.

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/79979).

odony commented:
> Shouldn't this be applied to v13 as well as the system parameters exists in that version?

Certainly yes, though at this point I'm mostly testing the idea ;)

It seems GMail does not do this all the time, but only when the mail went through an external domain (e.g. when it displays a `"via ..."` line), or when the mail is generally not trusted. In that case an optional config might be preferrable :thinking:

pedrobaeza commented:
I don't see bad to hide the email address also for privacy questions, so this should be if not always, at least by a configuration option.

pedrobaeza commented:
And also as strategy: showing your email tempts receivers to take that email and write directly to that one instead of replying and passing through Odoo hub.

xmo-odoo commented on odoo/tools/mail.py:
```diff
@@ -615,13 +615,14 @@ def formataddr(pair, charset='utf-8'):
     * Old From: "Admin" <admin@gmail.com>
     * New From: notifications@odoo.com
-    * Output:   "Admin (admin@gmail.com)" <notifications@odoo.com>
+    * Output:   "Admin (admin@gmail.com)" <notifications@odoo.com>  (if `include_email`)
+    *      or   "Admin" <notifications@odoo.com>  (it not `include_email`)
```
```suggestion
    * Output:   "Admin (admin@gmail.com)" <notifications@odoo.com>  (if ``include_email``)
    *      or   "Admin" <notifications@odoo.com>  (if not ``include_email``)
```

xmo-odoo commented on odoo/tools/mail.py:
```diff
@@ -631,10 +632,14 @@ def encapsulate_email(old_email, new_email):
+            name_part = old_name
     else:
-        name_part = old_email_split[0][1]
+        name_part = old_email
```
Isn't that going to trigger the same spoofing alert issue? You'll have an email as name part and a different email as actual email.

std-odoo replied:
I also think, it would be cool to replace `@` by `[at]` or something like this

odony replied:
We could replace the '@', but that may still look rather fishy, for smarter email providers. Maybe we just drop the domain and keep the local part? Better than nothing?

xmo-odoo commented on odoo/tools/mail.py:
```diff
@@ -615,13 +615,14 @@ def formataddr(pair, charset='utf-8'):
 
 
-def encapsulate_email(old_email, new_email):
+def encapsulate_email(old_email, new_email, include_email=False):
```
Could be kwonly.

Not sure about the name, but I don't really have a better one so far.

std-odoo replied:
is this parameter really useful ? (`include_email`) do we still want to include the email somewhere in the code ?

odony replied:
removed it, indeed we don't need to keep it

xmo-odoo commented on odoo/addons/base/tests/test_mail.py:
```diff
@@ -8,7 +8,7 @@
 import threading
 
-from odoo.tests.common import BaseCase, SavepointCase, TransactionCase
+from odoo.tests.common import BaseCase, SavepointCase, TransactionCase, tagged
```
Seems unused

std-odoo reviewed (commented):
Hi :) 

Good catch for this one :) what about still including the email but with `[at]` instead of `@` ? (so gmail might not detect the name as an email and remove it)

odony commented:
Thanks for the feedback! Removed the cruft, dropped the useless option to keep the old behavior (don't think it's useful indeed), and updated the commit msg/PR desc and tests.

I think we can proceed with this small fix if nobody has any blocking issue with it. It will definitely help decrease the risk of being detected as a spam or a tentative spoof by the various email service providers.

PS: For the 15.0 port, the logic for triggering the email rewrite is different (based on the `from_filter` of the mail server rather than only ICPs), but the code for the rewrite is the same, so the patch will apply cleanly in `tools.mail`. Of course the tests won't ;)

odony commented:
> what about still including the email but with `[at]` instead of `@` ? (so gmail might not detect the name as an email and remove it)

I chose to only keep the local part of the email instead, e.g if we only have an email `foo@bar.com`, then the rewritten name-email pair will become `"foo" <notifications@example.com>`.
I think it's safer than trying to disguise the `@` sign and being eventually detected as a spoofing attempt, in some future update of the secret heuristics of email service providers :)

odony commented:
@robodoo r+ please
